src/core/socket.c

```diff
@@ -468,13 +477,53 @@ static int socket_verify(Socket *s) {
         return 0;
 }
 
+static void socket_address_hash_func(const void *p, struct siphash *state) {
+        const SocketPeer *s = p;
+
+        assert(s);
+
+        siphash24_compress(&s->family, sizeof(s->family), state);
+
+        if (s->family == AF_INET)
+                siphash24_compress(&s->peer.in.sin_addr, sizeof(s->peer.in.sin_addr), state);
+        else if (s->family == AF_INET6)
+                siphash24_compress(&s->peer.in6.sin6_addr, sizeof(s->peer.in6.sin6_addr), state);
+}
```

Comment:
The hashfunc/comparison func really should just operate on on the sockaddr_union and not on the SocketPeer at all. The sockaddr contains everything we need after all: the address family as well as the actual address.